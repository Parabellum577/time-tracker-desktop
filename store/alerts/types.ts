import { ActionType } from 'typesafe-actions'

import * as actions from './actions'
import { IError } from '@services/types'

export type AlertsAction = ActionType<typeof actions>

export interface IAlert {
  alertType: 'success' | 'error' | 'warning' | 'coldstart'
  alertMessage: string | IError
}

export interface IAlertsState {
  alertsArray: IAlert[]
}
