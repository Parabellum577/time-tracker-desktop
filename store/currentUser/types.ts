import { ActionType } from 'typesafe-actions'

import { IUser, IDefaultUserSettings } from '@types'
import * as actions from './actions'

export type CurrentUserAction = ActionType<typeof actions>

export interface ICurrentUserState {
  readonly currentUser: IUser
  readonly defaultUserSettings: IDefaultUserSettings
}
