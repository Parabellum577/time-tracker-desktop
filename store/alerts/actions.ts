import { createAsyncAction } from 'typesafe-actions'
import { IError } from '@types'
import { IAlert } from './types'

export const showAlertMessage = createAsyncAction(
  'notifications/SHOW_ALERT_MESSAGE_REQUEST',
  'notifications/SHOW_ALERT_MESSAGE_SUCCESS',
  'notifications/SHOW_ALERT_MESSAGE_FAILURE',
)<IAlert, IAlert, IError>()

export const clearAlertMessage = createAsyncAction(
  'notifications/CLEAR_ALERT_MESSAGE_REQUEST',
  'notifications/CLEAR_ALERT_MESSAGE_SUCCESS',
  'notifications/CLEAR_ALERT_MESSAGE_FAILURE',
)<void, void, IError>()
