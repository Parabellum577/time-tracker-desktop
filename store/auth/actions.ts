import { createAction, createAsyncAction } from 'typesafe-actions'
import { IError } from '@types'
import { ILoginUserPayload, ILoginOTSPayload } from '@services/api-types'

export const loginUser = createAsyncAction('auth/LOGIN_REQUEST', 'auth/LOGIN_SUCCESS', 'auth/LOGIN_FAILURE')<
  ILoginUserPayload,
  void,
  IError
>()

export const loginUserViaOAuth = createAsyncAction(
  'auth/LOGIN_VIA_OAUTH_REQUEST',
  'auth/LOGIN_VIA_OAUTH_SUCCESS',
  'auth/LOGIN_VIA_OAUTH_FAILURE',
)<void, void, IError>()

export const otsLogin = createAsyncAction('auth/OTS_LOGIN_REQUEST', 'auth/OTS_LOGIN_SUCCESS', 'auth/OTS_LOGIN_FAILURE')<
  ILoginOTSPayload,
  ILoginOTSPayload,
  IError
>()

export const loadingDataAfterLogin = createAction('auth/LOADING_DATA_AFTER_LOGIN', resolve => (isLoading: boolean) =>
  resolve(isLoading),
)

export const signOut = createAction('auth/SIGN_OUT', resolve => () => resolve())
