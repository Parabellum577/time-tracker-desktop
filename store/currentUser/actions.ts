import { createAction, createAsyncAction } from 'typesafe-actions'
import { IUser, IError, IUpdateUser, ISettingsItem } from '@types'

export const getCurrentUser = createAsyncAction(
  'user/GET_CURRENT_USER_REQUEST',
  'user/GET_CURRENT_USER_SUCCESS',
  'user/GET_CURRENT_USER_FAILURE',
)<void, IUser, IError>()

export const updateCurrentUser = createAction('user/UPDATE_CURRENT_USER', action => (userData: IUpdateUser) =>
  action({
    Avatar: userData.Avatar,
    FirstName: userData.FirstName,
    LastName: userData.LastName,
    Phone: userData.Phone,
    Skype: userData.Skype,
    Role: userData.Role,
  }),
)

export const handleUserSettings = createAsyncAction(
  'user/HANDLE_USER_SETTINGS_REQUEST',
  'user/HANDLE_USER_SETTINGS_SUCCESS',
  'user/HANDLE_USER_SETTINGS_FAILURE',
)<ISettingsItem[], ISettingsItem[], IError>()

export const getUserSettings = createAsyncAction(
  'user/GET_USER_SETTINGS_REQUEST',
  'user/GET_USER_SETTINGS_SUCCESS',
  'user/GET_USER_SETTINGS_FAILURE',
)<void, void, IError>()

export const setDefaultUserSettings = createAction('user/SET_DEFAULT_USER_SETTINGS', action => action)

export const activateUserAccount = createAction('user/ACTIVATE_USER_ACCOUNT', action => action)
