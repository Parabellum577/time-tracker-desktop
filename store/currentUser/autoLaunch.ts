import persistStore from '../persistStore'
const Launch = require('auto-launch')

interface IConfig {
  name: string
  isHidden: boolean
  path?: string
}

const config: IConfig = {
  name: 'timetracker',
  isHidden: true,
}

if (process.platform === 'darwin') {
  config.path = `${process.execPath.match(/.*?\.app/)[0]}`
}

const AutoLaunchApp = new Launch(config)
if (!persistStore.get('isNotFirstStart')) {
  AutoLaunchApp.enable()
  persistStore.set('isNotFirstStart', true)
}

export const handleAutoLaunch = async () => {
  const isAutoLaunchOn = await AutoLaunchApp.isEnabled()
  if (isAutoLaunchOn) {
    await AutoLaunchApp.disable()
  } else {
    await AutoLaunchApp.enable()
  }
}

export const autoLaunchState = () => {
  return AutoLaunchApp.isEnabled()
}
