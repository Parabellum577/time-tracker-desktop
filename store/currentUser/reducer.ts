import { getType } from 'typesafe-actions'

import * as actions from './actions'
import { CurrentUserAction, ICurrentUserState } from './types'
import { IDefaultUserSettings } from '@types'

export const defaultSettings: IDefaultUserSettings = {
  isAppDefaultLaunch: 'off',
  isOverSpentMode: 'on',
  isAutoDailyReport: 'on',
  isNotifications: 'on',
  isSound: 'off',
  isInstantColdStart: 'off',
  isScreenshootNotifications: 'off',
  idle: '10',
  cold_settings: {
    TrackerID: 0,
    ProjectID: '',
    IssueID: '',
  },
  isDesktopInstalled: 'off',
  isShowedSearchBarTutorial: 'off',
  isShowedIntegrationTutorial: 'off',
  isShowedPersonalProjectTutorial: 'off',
  isShowedTaskAssignTutorial: 'off',
}

const initialState: ICurrentUserState = {
  currentUser: null,
  defaultUserSettings: defaultSettings,
}

interface IUserSettings {
  [type: string]: string
}

export default (state: ICurrentUserState = initialState, action: CurrentUserAction): ICurrentUserState => {
  switch (action.type) {
    case getType(actions.getCurrentUser.success):
      return { ...state, currentUser: action.payload }
    case getType(actions.getCurrentUser.failure):
      return { ...state }
    case getType(actions.updateCurrentUser):
      const { FirstName, LastName, Skype, Phone, Avatar, Role } = action.payload
      return {
        ...state,
        currentUser: { ...state.currentUser, FirstName, LastName, Skype, Phone, Avatar, Role },
      }
    case getType(actions.handleUserSettings.success):
      let settings = Object.values(action.payload)
      const isColdSettings = settings.some(s => s.Key === 'cold_settings')

      if (isColdSettings) {
        let coldSettings = settings.find(s => s.Key === 'cold_settings')
        coldSettings = { ...coldSettings, Value: JSON.parse(coldSettings.Value) }
        settings = [...settings, { ...coldSettings }]
      }

      const parsedUserSettings: IUserSettings = {}
      settings.map(setting => {
        parsedUserSettings[setting.Key] = setting.Value
      })

      return {
        ...state,
        defaultUserSettings: {
          ...state.defaultUserSettings,
          ...parsedUserSettings,
        },
      }

    case getType(actions.setDefaultUserSettings):
      return {
        ...state,
        defaultUserSettings: {
          ...defaultSettings,
        },
      }

    case getType(actions.handleUserSettings.failure):
      return { ...state }
    case getType(actions.activateUserAccount):
      return {
        ...state,
        currentUser: {
          ...state.currentUser,
          Activated: true,
        },
      }
    default:
      return state
  }
}
