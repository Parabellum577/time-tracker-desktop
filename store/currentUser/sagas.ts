import { getAvatar } from '@services/microservices/avatar'
import { SagaIterator, Task } from 'redux-saga'
import { call, put, take, race, delay, select, fork, cancel, putResolve } from 'redux-saga/effects'
import { getType } from 'typesafe-actions'
import * as _ from 'lodash'

import * as actions from './actions'
import * as loginActions from '../auth/actions'
import api from '@api'
import { IUser, IUserSettings, ISettingsItem, IDefaultUserSettings } from '@types'
import { IRootState } from '@store/rootReducer'
import { ISettingsItemPayload } from '@api-types'
import { handleAutoLaunch, autoLaunchState } from './autoLaunch'
import { showNotification } from '@store/notifications/actions'

export function* getCurrentUser(): SagaIterator {
  try {
    const userData = yield call(api.user.GetCurrentUser)
    const userAvatar = yield call(getAvatar)
    const user: IUser = { ...userData, Avatar: userAvatar }
    yield put(actions.getCurrentUser.success(user))
  } catch (e) {
    yield put(actions.getCurrentUser.failure(e))
  }
}

export function* getCurrentUserSaga() {
  while (true) {
    const { signOut } = yield race({
      getCurrentUserRequest: take(getType(actions.getCurrentUser.request)),
      signOut: take(getType(loginActions.signOut)),
      userLogin: take(getType(loginActions.loginUser.success)),
    })
    if (signOut) {
      yield put(actions.getCurrentUser.success(null))
    } else {
      yield call(getCurrentUser)
    }
  }
}

export const getUserSettingsFromStore = (state: IRootState) => state.user.defaultUserSettings

export function* getUserSettingsSaga(): SagaIterator {
  while (true) {
    const { userSettingChanged } = yield race({
      userSettingChanged: take(getType(actions.handleUserSettings.request)),
      userLogin: take(getType(loginActions.loginUser.success)),
    })

    if (userSettingChanged) {
      const newSetting: ISettingsItem[] = userSettingChanged.payload
      yield call(handleUserSettings, newSetting)
    } else {
      try {
        const defaultSettings: ISettingsItem[] = yield call(api.user.GetUserSettings)
        yield call(getUserSettings, defaultSettings)
        yield put(actions.getUserSettings.success())
      } catch (error) {
        yield put(actions.getUserSettings.failure(error))
      }
    }
  }
}

let settingsChangesTask: Task = null
let settingsChangesQueue: ISettingsItemPayload[] = []

export function* handleUserSettings(newSettings: ISettingsItem[]) {
  try {
    for (const setting of newSettings) {
      const Key = setting.Key
      let newPayload: ISettingsItemPayload[] = []
      const settings = yield select(getUserSettingsFromStore)
      const storeSettingsKeys = Object.keys(settings)
      const oldSettings = settings[storeSettingsKeys.find(settingKey => settingKey === Key)]

      if (Key === 'isAppDefaultLaunch') {
        const isColdStart = settings.isInstantColdStart === 'on'
        yield call(handleAutoLaunch)
        const currentAutoLaunch = yield call(autoLaunchState)
        newPayload = [{ Key, Value: currentAutoLaunch ? 'on' : 'off' }]
        if (isColdStart && !currentAutoLaunch) {
          newPayload = [...newPayload, { Key: 'isInstantColdStart', Value: 'off' }]
        }
      } else {
        if (Key === 'idle') {
          const Value = setting.Value
          const TrackerID = settings.cold_settings.TrackerID
          const ProjectID = settings.cold_settings.ProjectID
          newPayload = [{ Key, Value }]
          yield call(api.project.SetSettings, {
            TrackerID,
            ProjectID,
            Value,
            Key: 'track-idle-detection',
          })
        } else {
          if (Key === 'isInstantColdStart') {
            const currentAutoLaunch = yield call(autoLaunchState)
            if (!currentAutoLaunch) {
              yield call(handleAutoLaunch)
              newPayload = [{ Key: 'isAppDefaultLaunch', Value: 'on' }]
            }
          }
          if (Key === 'isScreenshootNotifications') {
            oldSettings === 'on'
              ? yield put(showNotification.request({ notificationType: 'ScreenshottingOff' }))
              : yield put(showNotification.request({ notificationType: 'ScreenshottingOn' }))
          }

          const tutorialKeys = [
            'isDesktopInstalled',
            'isShowedSearchBarTutorial',
            'isShowedIntegrationTutorial',
            'isShowedPersonalProjectTutorial',
            'isShowedTaskAssignTutorial',
          ]

          if (tutorialKeys.includes(Key)) {
            newPayload = [...newPayload, { Key, Value: setting.Value }]
          }

          newPayload = [...newPayload, { Key, Value: oldSettings === 'on' ? 'off' : 'on' }]
        }
      }

      const isAlreadyExist = settingsChangesQueue.some(config => newPayload.some(p => p.Key === config.Key))

      if (isAlreadyExist) {
        settingsChangesQueue = [
          ...settingsChangesQueue.filter(config => !newPayload.find(p => p.Key === config.Key)),
          ...newPayload,
        ]
      } else {
        settingsChangesQueue = [...settingsChangesQueue, ...newPayload]
      }

      if (!settingsChangesTask) {
        settingsChangesTask = yield fork(settingsQueueResolveSaga)
      }

      yield put(actions.handleUserSettings.success(newPayload))
    }
  } catch (error) {
    console.error(error)
  }
}

export function* settingsQueueResolveSaga(): SagaIterator {
  while (true) {
    if (settingsChangesQueue.length > 0) {
      const newConfig: ISettingsItemPayload = settingsChangesQueue[0]
      settingsChangesQueue = settingsChangesQueue.slice(1)
      try {
        yield call(api.user.SetUserSettings, newConfig)
      } catch (error) {
        console.error(error)
      }
      yield delay(500)
    } else {
      settingsChangesTask = null
      break
    }
  }
}

export function* setDefaultUserSettings(): SagaIterator {
  while (true) {
    yield take(getType(loginActions.signOut))
    yield put(actions.setDefaultUserSettings())
  }
}

export function* getUserSettings(defaultSettings: ISettingsItem[]) {
  try {
    const settings: IDefaultUserSettings = yield select(getUserSettingsFromStore)
    const storeSettingsKeys = Object.keys(settings)
    const optionValues = Object.values(defaultSettings)
    const savedUserSettings = optionValues.filter(f => storeSettingsKeys.includes(f.Key))
    const currentAutoLaunch = yield call(autoLaunchState)
    const isHaveOverspentSetting = settings.isAutoPauseTask

    yield putResolve(
      actions.handleUserSettings.success([
        ...savedUserSettings,
        { Key: 'isAppDefaultLaunch', Value: currentAutoLaunch ? 'on' : 'off' },
      ]),
    )

    if (isHaveOverspentSetting) {
      yield put(
        actions.handleUserSettings.success([
          ...savedUserSettings,
          { Key: 'isOverSpentMode', Value: settings.isAutoPauseTask === 'on' ? 'off' : 'on' },
        ]),
      )
    }
  } catch (error) {
    yield put(actions.handleUserSettings.failure(error))
    console.error(error)
  }
}
