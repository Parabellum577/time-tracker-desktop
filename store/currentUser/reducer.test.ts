import { ActionType, getType } from 'typesafe-actions'

import * as actions from './actions'
import { ICurrentUserState } from './types'
import currentUser, { defaultSettings } from './reducer'

import { SIMPLE_USER } from '@test-helpers/sample'

describe('[Store] CurrentUser Reducer', () => {
  it('[Success] [Snapshot] Should handle getCurrentUser', () => {
    const action: ActionType<typeof actions> = {
      type: getType(actions.getCurrentUser.success),
      payload: { ...SIMPLE_USER },
    }

    const currentState: ICurrentUserState = {
      currentUser: { ...SIMPLE_USER },
      defaultUserSettings: defaultSettings,
    }

    expect(currentUser(currentState, action)).toMatchSnapshot()
  })

  it('[Failure] Should handle getCurrentUser', () => {
    const action: ActionType<typeof actions> = {
      type: getType(actions.getCurrentUser.failure),
      payload: {
        code: 666,
        message: 'SOME_ERROR',
      },
    }

    const currentState: ICurrentUserState = {
      currentUser: null,
      defaultUserSettings: defaultSettings,
    }

    expect(currentUser(currentState, action)).toMatchSnapshot()
  })

  it('[Success] [Snapshot] Should handle updateCurrentUser', () => {
    const action: ActionType<typeof actions> = {
      type: getType(actions.updateCurrentUser),
      payload: { ...SIMPLE_USER },
    }

    const currentState: ICurrentUserState = {
      currentUser: { ...SIMPLE_USER },
      defaultUserSettings: defaultSettings,
    }

    expect(currentUser(currentState, action)).toMatchSnapshot()
  })
})
