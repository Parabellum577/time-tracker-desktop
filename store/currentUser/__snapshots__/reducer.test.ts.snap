// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Store] CurrentUser Reducer [Failure] Should handle getCurrentUser 1`] = `
Object {
  "currentUser": null,
  "defaultUserSettings": Object {
    "cold_settings": Object {
      "IssueID": "",
      "ProjectID": "",
      "TrackerID": 0,
    },
    "idle": "10",
    "isAppDefaultLaunch": "off",
    "isAutoDailyReport": "on",
    "isDesktopInstalled": "off",
    "isInstantColdStart": "off",
    "isNotifications": "on",
    "isOverSpentMode": "off",
    "isScreenshootNotifications": "off",
    "isShowedIntegrationTutorial": "off",
    "isShowedPersonalProjectTutorial": "off",
    "isShowedSearchBarTutorial": "off",
    "isShowedTaskAssignTutorial": "off",
    "isSound": "off",
  },
}
`;

exports[`[Store] CurrentUser Reducer [Success] [Snapshot] Should handle getCurrentUser 1`] = `
Object {
  "currentUser": Object {
    "Activated": true,
    "Avatar": "",
    "ConfirmationSent": 0,
    "Deleted": false,
    "FirstName": "Sam",
    "IsAdmin": false,
    "IsAnonymous": false,
    "LastName": "Smith",
    "Login": "test@gmail.com",
    "PasswordUpdatedAt": 0,
    "Phone": "911",
    "Registered": 1538382878,
    "Role": "Other",
    "Skype": "whatIsIt",
    "SubscribeOnMail": false,
    "Timezone": 10800,
    "UpdatedAt": 1555054724,
    "UserID": 27311,
  },
  "defaultUserSettings": Object {
    "cold_settings": Object {
      "IssueID": "",
      "ProjectID": "",
      "TrackerID": 0,
    },
    "idle": "10",
    "isAppDefaultLaunch": "off",
    "isAutoDailyReport": "on",
    "isDesktopInstalled": "off",
    "isInstantColdStart": "off",
    "isNotifications": "on",
    "isOverSpentMode": "off",
    "isScreenshootNotifications": "off",
    "isShowedIntegrationTutorial": "off",
    "isShowedPersonalProjectTutorial": "off",
    "isShowedSearchBarTutorial": "off",
    "isShowedTaskAssignTutorial": "off",
    "isSound": "off",
  },
}
`;

exports[`[Store] CurrentUser Reducer [Success] [Snapshot] Should handle updateCurrentUser 1`] = `
Object {
  "currentUser": Object {
    "Activated": true,
    "Avatar": "",
    "ConfirmationSent": 0,
    "Deleted": false,
    "FirstName": "Sam",
    "IsAdmin": false,
    "IsAnonymous": false,
    "LastName": "Smith",
    "Login": "test@gmail.com",
    "PasswordUpdatedAt": 0,
    "Phone": "911",
    "Registered": 1538382878,
    "Role": "Other",
    "Skype": "whatIsIt",
    "SubscribeOnMail": false,
    "Timezone": 10800,
    "UpdatedAt": 1555054724,
    "UserID": 27311,
  },
  "defaultUserSettings": Object {
    "cold_settings": Object {
      "IssueID": "",
      "ProjectID": "",
      "TrackerID": 0,
    },
    "idle": "10",
    "isAppDefaultLaunch": "off",
    "isAutoDailyReport": "on",
    "isDesktopInstalled": "off",
    "isInstantColdStart": "off",
    "isNotifications": "on",
    "isOverSpentMode": "off",
    "isScreenshootNotifications": "off",
    "isShowedIntegrationTutorial": "off",
    "isShowedPersonalProjectTutorial": "off",
    "isShowedSearchBarTutorial": "off",
    "isShowedTaskAssignTutorial": "off",
    "isSound": "off",
  },
}
`;
